[Clang][NFC] Rename UnqualPtrTy to DefaultPtrTy

jmmartinez opened the pull request:
`UnqualPtrTy` didn't always match `llvm::PointerType::getUnqual`: sometimes it returned a pointer that is not in address space 0 (notably for SPIRV).

Since `UnqualPtrTy` was used as the "generic" or "default" pointer type, this patch renames it to `DefaultPtrTy` to avoid confusion with LLVM's `PointerType::getUnqual`.

efriedma-quic reviewed (commented):
How do you actually plan to use this?  In almost all cases, clang is not supposed to be hardcoding address-space zero.  (Most existing places are just code nobody has gotten around to updating yet.)

efriedma-quic commented:
I think on the rare occasion where we actually want addrspace 0, regardless of the target's addrspace preferences, I'd prefer to change it to explicitly request addrspace 0.  `Builder.getPtrTy(/*AddrSpace*/0)` is unambiguous.

arsenm commented:
> So on LLVM side it is (often) assumed that the "default address space" always is AS=0. But on clang side targets can map DefaultAS to a non zero address space. I wonder if that inconsistency is a bigger problem than addressed by this patch.

Yes, this is a hack from the beginning of OpenCL. Instead of changing the default address space, it would be better if clang treated OpenCL 1.x as implicitly adding the private qualifier on pointers

jmmartinez committed:
[Clang][NFC] Rename UnqualPtrTy to DefaultPtrTy

UnqualPtrTy didn't always match llvm::PointerType::getUnqual:
sometimes it returned a pointer that is not in addrspace 0 (notably for
SPIRV).

Since UnqualPtrTy was used as the "generic" or "default" pointer type,
this patch renames it to DefaultPtrTy to avoid confusion with LLVM's
PointerType::getUnqual.

efriedma-quic reviewed (approved):
LGTM

bjope commented on clang/lib/CodeGen/ItaniumCXXABI.cpp:
```diff
@@ -816,7 +816,7 @@ CGCallee ItaniumCXXABI::EmitLoadOfMemberFunctionPointer(
   CGF.EmitBlock(FnNonVirtual);
   llvm::Value *NonVirtualFn =
-      Builder.CreateIntToPtr(FnAsInt, CGF.UnqualPtrTy, "memptr.nonvirtualfn");
+      Builder.CreateIntToPtr(FnAsInt, CGF.DefaultPtrTy, "memptr.nonvirtualfn");
```
I guess this is an example when it gets a bit weird.

Downstream we actually have a special CGF.FnVoidPtrTy that is used here. As we want function pointers to use the ProgramAddressSpace from the DataLayout (the default address space for functions is not necessarily the same as for data). That is however not implemented in the upstream clang.

Using CGF.DefaultPtrTy here (being a pointer to the default data memory) seems a bit weird for Harvard Architectures when the program memory is different from the data memory.

So one could wonder if UnqualPtrTy was used here to denote that these function pointers should be treated as unqualified entities without having any address space associated with them (although then one could wonder how one know the size of the pointers)? Or at least to indicate that this is a pointer to possibly something else than just the default data address space.

(With that being said I'm thinking that maybe this refactoring is OK anyway, since it is only a kind of churn that rename things. Downstream we will have to deal with conflicts due to the fact that we actually identify function pointers a bit differently. Maybe that is something that we should consider upstreaming.)

jmmartinez commented:
Thanks for your help !
